lib/puppet/ssl/state_machine.rb

```diff
@@ -260,10 +263,16 @@ def ensure_client_certificate
   private
 
   def run_machine(state, stop)
-    loop do
-      state = state.next_state
+    lock do
+      loop do
+        state = state.next_state
 
-      return state if state.is_a?(stop)
+        break if state.is_a?(stop)
+      end
     end
+
+    state
+  rescue Puppet::LockError
+    raise Puppet::Error, 'Another puppet instance is already running; exiting'
```

Comment:
Do we use this locking elsewhere?  Is it useful to the user to have some detail about why this exclusion is happening?

Reply:
Yes the configurer (used by both agent and apply) will acquire the lock prior to getting/applying a catalog.